snuba/datasets/entities/sessions.py

```diff
@@ -250,3 +257,59 @@ def get_query_processors(self) -> Sequence[QueryProcessor]:
                 {"bucketed_started": "started"}, ("started", "received")
             ),
         ]
+
+
+GRANULARITY_MAPPING = {
+    "toStartOfMinute": 60,
+    "toStartOfHour": 3600,
+    "toDate": 86400,
+}
+
+
+def extract_granularity_from_query(query: Query) -> Optional[int]:
+    """
+    This extracts the `granularity` from the `groupby` statement of the query,
+    if the query is grouped by the `bucketed_started` virtual column
+    (which is a transformation of the `started` column).
+    The matches are essentially the reverse of `TimeSeriesProcessor.__group_time_function`.
+    """
+    groupby = query.get_groupby_from_ast()
+
+    column_match = ColumnMatch(None, String("started"))
+    started_fn_match = FunctionCallMatch(
+        Param(
+            "time_fn",
+            Or([String("toStartOfHour"), String("toStartOfMinute"), String("toDate")]),
+        ),
+        (column_match, LiteralMatch(Any(str))),
+    )
+    started_expr_match = FunctionCallMatch(
+        String("toDateTime"),
+        (
+            FunctionCallMatch(
+                String("multiply"),
+                (
+                    FunctionCallMatch(
+                        String("intDiv"),
+                        (
+                            FunctionCallMatch(String("toUInt32"), (column_match,)),
+                            LiteralMatch(Param("granularity", Any(int))),
+                        ),
+                    ),
+                    LiteralMatch(Param("granularity", Any(int))),
+                ),
+            ),
+            LiteralMatch(Any(str)),
+        ),
+    )
+
+    for expr in groupby:
```

Comment:
It would be safer to run it this way:
```
for expr in groupby:
     for sub_expr in expr:
          # do the rest as you wrote.
```

Every `Expression` is an iterable that traverses every node of the expression tree. So technically if somebody has a `ifNotNullDoSomethingAboutMytime(toStartOfHour(start_date))`, by doing the nested loop you would still be able to apply your pattern to the inner `toStartOfHour(start_date)`.